### PR TITLE
Manually type-cast all AJAX files

### LINKED ISSUE #14196: Use type casting in AJAX files

### PR DESCRIPTION
**Pull request by LukasReschke**
This enforces proper types on POST and GET arguments where I considered it sensible. I didn't update some as I don't know what kind of values they would support :see_no_evil:

Proper solution would be to use the AppFramework instead such static files as the AppFramework supports this type casting already using annotations…

Fixes https://github.com/owncloud/core/issues/14196 for core – master only.

### CODE REVIEW

**Comment by LukasReschke** [comment]
@nickvergessen @th3fallen @MorrisJobke Review fun? :see_no_evil:

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

**Comment by LukasReschke** [comment]
Please notice that this has actually very little impact as usual $_POST or $_GET argument anyways is a `string`, this just prevents calling them as array which can lead to bad unexpected side-effects.

There is in some places a cast to `bool` or `int` where I knew that only this are accepted values.

File: apps/files/ajax/delete.php [hunk 1]
```diff
@@ -6,18 +6,18 @@
 
 
 // Get data
-$dir = isset($_POST['dir']) ? $_POST['dir'] : '';
-$allFiles = isset($_POST["allfiles"]) ? $_POST["allfiles"] : false;
+$dir = isset($_POST['dir']) ? (string)$_POST['dir'] : '';
+$allFiles = isset($_POST["allfiles"]) ? (bool)$_POST["allfiles"] : false;
```
Comment (nickvergessen) [inline apps/files/ajax/delete.php]: behaviour change, string `'false'`  did not run `$allFiles` previously, now it does?

Comment (LukasReschke) [inline apps/files/ajax/delete.php]: If `false` it will now evaluate to `false` and not get into L13

Comment (LukasReschke) [inline apps/files/ajax/delete.php]: Adjusting. You're right :see_no_evil:

Comment (nickvergessen) [inline apps/files/ajax/delete.php]: fixed

File: apps/files_versions/ajax/getVersions.php [hunk 2]
```diff
@@ -3,8 +3,8 @@
 OCP\JSON::callCheck();
 OCP\JSON::checkAppEnabled('files_versions');
 
-$source = $_GET['source'];
-$start = $_GET['start'];
+$source = (string)$_GET['source'];
+$start = (string)$_GET['start'];
```
Comment (PVince81) [inline apps/files_versions/ajax/getVersions.php]: int ?

Comment (LukasReschke) [inline apps/files_versions/ajax/getVersions.php]: Makes sense. Though would also work that way.

File: apps/files_encryption/ajax/userrecovery.php [hunk 3]
```diff
@@ -23,7 +23,7 @@
 	$util = new \OCA\Files_Encryption\Util($view, $userId);
 
 	// Save recovery preference to DB
-	$return = $util->setRecoveryForUser($_POST['userEnableRecovery']);
+	$return = $util->setRecoveryForUser((string)$_POST['userEnableRecovery']);
```
Comment (nickvergessen) [inline apps/files_encryption/ajax/userrecovery.php]: is this bool/int ?

Comment (LukasReschke) [inline apps/files_encryption/ajax/userrecovery.php]: Some blackmagic that I didn't dare to touch:

```php
/**
 * Enable / disable pwd recovery for a given user
 * @param bool $enabled Whether to enable or disable recovery
 * @return bool
 */
public function setRecoveryForUser($enabled) {

  $value = $enabled ? '1' : '0';
  try {
    \OC::$server->getConfig()->setUserValue($this->userId, 'files_encryption', 'recovery_enabled', $value);
    return true;
  } catch(\OCP\PreConditionNotMetException $e) {
    return false;
  }

}
```

Comment (LukasReschke) [inline apps/files_encryption/ajax/userrecovery.php]: as string should always work anyways :)

Comment (nickvergessen) [inline apps/files_encryption/ajax/userrecovery.php]: ok

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by MorrisJobke** [comment]
Just tested the usual stuff ... all works :+1:

**Comment by LukasReschke** [comment]
Thanks guys! :smile:

**Pull request merged**